native-agent/src/env.rs

```diff
@@ -0,0 +1,215 @@
+        })
+    }
+
+    pub fn wrap(jvmti_env: *mut jvmtiEnv) -> JvmTiEnv {
+        JvmTiEnv { jvmti: jvmti_env }
+    }
+
+    pub fn enable_capabilities(&mut self) -> Result<()> {
+        let mut capabilities = jvmtiCapabilities::default();
+        capabilities.set_can_generate_exception_events(1u32);
+        capabilities.set_can_access_local_variables(1u32);
+        jvmtifn!(self.jvmti, AddCapabilities, &capabilities)
+    }
+
+    pub fn set_exception_handler(&mut self) -> Result<()> {
+        let callbacks = jvmtiEventCallbacks {
+            Exception: Some(c_on_exception),
+            ..Default::default()
+        };
+
+        // This binding is necessary, the type checker can't handle ? here
+        let a: Result<()> = jvmtifn!(
+            self.jvmti,
+            SetEventCallbacks,
+            &callbacks,
+            size_of::<jvmtiEventCallbacks>() as i32
+        );
+        if a.is_err() {
+            return a;
+        }
+        jvmtifn!(
+            self.jvmti,
+            SetEventNotificationMode,
+            jvmtiEventMode_JVMTI_ENABLE,
+            jvmtiEvent_JVMTI_EVENT_EXCEPTION,
+            ptr::null_mut()
+        )
+    }
+
+    pub fn get_frame_count(&mut self, thread: jthread) -> Result<jint> {
+        let mut result: jint = 0;
+        let rc;
+        unsafe {
+            rc = (**self.jvmti)
+                .GetFrameCount
+                .expect("GetFrameCount function not found")(
+                self.jvmti, thread, &mut result
+            );
+        }
+        if rc != jvmtiError_JVMTI_ERROR_NONE {
+            let message = "JVMTI GetFrameCount failed".to_owned();
+            bail!(ErrorKind::JvmTi(message, rc as i32))
+        } else {
+            Ok(result)
+        }
+    }
+
+    pub fn get_stack_trace(
+        &mut self,
+        thread: jthread,
+        start_depth: jint,
+        num_frames: jint,
+        frame_buffer: *mut jvmtiFrameInfo,
+        count_ptr: *mut jint,
+    ) -> Result<()> {
+        jvmtifn!(
+            self.jvmti,
+            GetStackTrace,
+            thread,
+            start_depth,
+            num_frames,
+            frame_buffer,
+            count_ptr
+        )
+    }
+
+    pub fn get_local_variable_table(
+        &mut self,
+        method: jmethodID,
+        num_entries: &mut jint,
+        local_var_table: *mut *mut jvmtiLocalVariableEntry,
+    ) -> Result<()> {
+        jvmtifn!(
+            self.jvmti,
+            GetLocalVariableTable,
+            method,
+            num_entries,
+            local_var_table
+        )
+    }
+
+    pub fn dealloc<T>(&mut self, ptr: *mut T) -> Result<()> {
+        jvmtifn!(self.jvmti, Deallocate, ptr as *mut c_uchar)
+    }
+
+    pub fn get_local_object(
+        &mut self,
+        thread: jthread,
+        depth: jint,
+        slot: jint,
+        result: *mut jobject,
+    ) -> Result<()> {
+        jvmtifn!(self.jvmti, GetLocalObject, thread, depth, slot, result)
```

Comment:
This is the only part that I'm not 100% on. The docs say the object returned in the result pointer here is a JNI local reference which must be "managed". I am pretty sure we are in the clear here, but if there were to be a problem it would come from this.